hazelcast/src/main/java/com/hazelcast/util/collection/InflatableSet.java

```diff
@@ -0,0 +1,223 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.util.collection;
+
+import java.util.AbstractSet;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Provides fast {@link Set} implementation for cases where items are known to not
+ * contain duplicates.
+ *
+ * It doesn't not call equals/hash methods on initial data insertion hence it avoids
+ * performance penalty in the case these methods are expensive. It also means it does
+ * not detect duplicates - it's a responsibility of a caller to make sure no duplicated
+ * entries are inserted.
+ *
+ * Once the initial load is done then caller should indicate it
+ * by calling {@link #close()}. It switches operational mode and if any item is
+ * inserted afterward then the Set will inflate - it copies its content into
+ * internal HashSet. This means duplicate detection is enabled, but obviously
+ * it ruins the initial performance gain. We are making a bet the Set will not be
+ * modified once it's closed. The Set will also inflate on {@link #contains(Object)}
+ * and {@link #containsAll(Collection)} calls to enable fast look-ups.
+ *
+ * It's intended to be use in cases where a contract mandates us to return Set,
+ * but we know our data contains not duplicates. It performs the best in cases
+ * biased towards sequential iterations.
+ *
+ *
+ * @param <T>
+ */
+public class InflatableSet<T> extends AbstractSet<T> implements Set<T> {
+    private enum State {
+        //Set is still open for initial load. It's caller's responsibility to make
+        //sure no duplicates are added to this Set in this state
+        INITIAL_LOAD,
+
+        //Only array-backed representation exists
+        COMPACT,
+
+        //both array-backed & hashset-backed representation exist.
+        //this is needed as we are creating HashSet on contains()
+        //but we don't want invalidate existing iterators.
+        HYBRID,
+
+        //only hashset based representation exists
+        INFLATED
+    }
+
+    private final List<T> compactList;
+    private Set<T> inflatedSet;
+    private State state = State.INITIAL_LOAD;
+
+    public InflatableSet(int initialCapacity) {
+        if (initialCapacity < 0) {
+            throw new IllegalArgumentException("Initial Capacity must be a positive number. Current capacity: "
+                    + initialCapacity);
+        }
+        compactList = new ArrayList<T>(initialCapacity);
```

Comment:
I can not see a disadvantage but if you have a scenario which requires `add` and `remove` combinations in INITIAL_LOAD state, it will perform better.